Mathlib/GroupTheory/Perm/Centralizer.lean

```diff
@@ -0,0 +1,642 @@
+  one_mem' := by
+    simp only [Set.mem_setOf_eq, coe_one, id_eq, eq_self_iff_true, imp_true_iff]
+  mul_mem' hσ hτ := by
+    simp only [Subtype.forall, Set.mem_setOf_eq, coe_mul, Function.comp_apply]
+    simp only [Subtype.forall, Set.mem_setOf_eq] at hσ hτ
+    intro c hc
+    rw [hσ, hτ]
+  inv_mem' hσ := by
+    simp only [Subtype.forall, Set.mem_setOf_eq] at hσ ⊢
+    intro c hc
+    rw [← hσ]
+    · simp only [Finset.coe_mem, Subtype.coe_eta, apply_inv_self]
+    · simp only [Finset.coe_mem]
+
+variable {g} in
+theorem mem_range_toPermHom'_iff {τ : Perm g.cycleFactorsFinset} :
+    τ ∈ range_toPermHom' g ↔
+      ∀ c, (τ c : Perm α).support.card = (c : Perm α).support.card :=
+  Iff.rfl
+
+variable (k : centralizer {g})
+
+/-- `k : Subgroup.centralizer {g}` belongs to the kernel of `toPermHom g`
+  iff it commutes with each cycle of `g` -/
+theorem mem_ker_toPermHom_iff :
+    k ∈ (toPermHom g).ker ↔ ∀ c ∈ g.cycleFactorsFinset, Commute (k : Perm α) c := by
+  simp only [toPermHom, MonoidHom.mem_ker, DFunLike.ext_iff, Subtype.forall]
+  refine forall₂_congr (fun _ _ ↦ ?_)
+  simp [← Subtype.coe_inj, commute_iff_eq, mul_inv_eq_iff_eq_mul]
+
+end OnCycleFactors
+
+open OnCycleFactors Subgroup
+
+/-- A `Basis` of a permutation is a choice of an element in each of its cycles -/
+structure Basis (g : Equiv.Perm α) where
+  /-- A choice of elements in each cycle -/
+  (toFun : g.cycleFactorsFinset → α)
+  /-- For each cycle, the chosen element belongs to the cycle -/
+  (mem_support_self' : ∀ (c : g.cycleFactorsFinset), toFun c ∈ (c : Perm α).support)
+
+instance (g : Perm α) :
+    DFunLike (Basis g) (g.cycleFactorsFinset) (fun _ => α) where
+  coe a := a.toFun
+  coe_injective' a a' _ := by cases a; cases a'; congr
+
+namespace Basis
+
+theorem nonempty (g : Perm α) : Nonempty (Basis g) := by
+  have (c : g.cycleFactorsFinset) : (c : Perm α).support.Nonempty :=
+    IsCycle.nonempty_support (mem_cycleFactorsFinset_iff.mp c.prop).1
+  exact ⟨fun c ↦ (this c).choose, fun c ↦ (this c).choose_spec⟩
+
+variable (a : Basis g) (c : g.cycleFactorsFinset)
+
+theorem mem_support_self :
+    a c ∈ (c : Perm α).support := a.mem_support_self' c
+
+theorem injective : Function.Injective a := by
+  intro c d h
+  rw [← Subtype.coe_inj]
+  apply g.cycleFactorsFinset_pairwise_disjoint.eq c.prop d.prop
+  simp only [Disjoint, not_forall, not_or]
+  use a c
+  conv_rhs => rw [h]
+  simp only [← Perm.mem_support, a.mem_support_self c, a.mem_support_self d, and_self]
+
+theorem cycleOf_eq : g.cycleOf (a c) = c :=
+  (cycle_is_cycleOf (a.mem_support_self c) c.prop).symm
+
+theorem sameCycle {x : α} (hx : g.cycleOf x ∈ g.cycleFactorsFinset) :
+    g.SameCycle (a ⟨g.cycleOf x, hx⟩) x :=
+  (mem_support_cycleOf_iff.mp (a.mem_support_self ⟨g.cycleOf x, hx⟩)).1.symm
+
+variable (τ : range_toPermHom' g)
+
+/-- The function that will provide a right inverse `toCentralizer` to `toPermHom` -/
+def newK (x : α) : α :=
```

Comment:
Is this a temporary definition that you only need in this file? Or do you want it available outside this file as well?

If it is only for this file, please consider marking it as `private def` (and the same for all results mentioning it in their statement).

Reply:
I believe we should have it outside of the file, because it presents the centralizer as a semidirect product.
Hence it should have a better name.

Reply:
But it will remain available as `a.ofPermHom τ` of which you prove that it lies in the centralizer. Is that sufficient? One option could be to make `newK` private, and make `a.ofPermHom τ` the official spelling outside this file. But then all the API that people need will have to be made available about `a.ofPermHom τ`.
Of course such API can be added in later PRs.

I am inclined to go this route. If you agree, then I suggest to add to the docstring of `newK` that it is an implementation detail of `a.ofPermHom τ`.

Reply:
I renamed it as `Equiv.Perm.Basis.ofPermHom_fun`.
If I make it private, then it is logical to make private all the lemmas that use it explicitly, and one would have to repeat a lot of them for `Equiv.Perm.Basis.ofPermHom`.